examples/chat/module/rust/src/lib.rs

```diff
@@ -14,137 +14,92 @@
-        info!("creating room");
-        if self.rooms.contains_key(&req.room_id) {
-            return room_id_duplicate_err();
-        }
-
-        // Create a new Node for this room, and keep the write handle and admin token in the `rooms`
-        // map.
-        self.rooms.insert(req.room_id, Room::new(req.admin_token));
-
-        Ok(Empty::new())
+// TODO: Make this node generic over a trait `Routable` or similar, which determines what to route
+// incoming messages based on.
+impl Node<Invocation> for Router {
+    fn new() -> Self {
+        Router::default()
     }
 
-    fn destroy_room(&mut self, req: DestroyRoomRequest) -> grpc::Result<Empty> {
-        info!("destroying room");
-        match self.rooms.entry(req.room_id) {
-            Entry::Occupied(e) => {
-                if e.get().admin_token == req.admin_token {
-                    // Close the only input channel that reaches the per-room Node, which
-                    // will trigger it to terminate.
-                    e.get().sender.close().expect("could not close channel");
-                    e.remove();
-                    Ok(Empty::new())
-                } else {
-                    Err(grpc::build_status(
-                        grpc::Code::PERMISSION_DENIED,
-                        "invalid admin token",
-                    ))
-                }
-            }
-            Entry::Vacant(_) => room_id_not_found_err(),
+    fn handle_command(&mut self, grpc_invocation: Invocation) -> Result<(), oak::OakError> {
+        let grpc_request_label = grpc_invocation.request_receiver.label()?;
+        let grpc_response_label = grpc_invocation.response_sender.label()?;
+        info!(
+            "handling incoming gRPC invocation with labels: request={:?} response={:?}",
+            grpc_request_label, grpc_response_label
+        );
+        if grpc_request_label.can_flow_to(&grpc_response_label) {
+            // Retrieve a reference to a room for the request label, or create one if it does not
+            // already exist.
+            let room_node_sender = self
+                .rooms
+                .entry(grpc_request_label.clone())
+                .or_insert_with(|| create_room_node(&grpc_request_label));
+            // Forward the incoming gRPC invocation to the appropriate room node.
+            room_node_sender.send(&grpc_invocation)
+        } else {
+            // If the caller is trying to send a message with a label such that it would not be able
+            // to receive it itself, do not even attempt to process the request and instead return
+            // an error immediately.
+            warn!("gRPC request and response have inconsistent labels");
+            let mut grpc_response_writer =
+                oak::grpc::ChannelResponseWriter::new(grpc_invocation.response_sender);
+            grpc_response_writer.close(Err(oak::grpc::build_status(
+                oak::grpc::Code::PERMISSION_DENIED,
```

Comment:
Is this propagated to the client? In that case, it is possible to leak data. As a general rule, all failures should be silent unless in debug mode.